File: servers/python/satellite_server/protocol/pull.py
"""Pull operation for the Satellite sync protocol."""

from __future__ import annotations

import json

from satellite_server.interfaces import IObjectStore
from satellite_server.timestamp import next_timestamp
from satellite_server.protocol.types import StoredDocument, PullResult
from satellite_server.protocol.timestamps import filter_by_checkpoint


async def pull(
    store: IObjectStore,
    document_key: str,
    checkpoint: int = 0,
) -> PullResult:
    """Pull the current document, optionally filtered by checkpoint.

    - No checkpoint (or 0): returns full data
    - With checkpoint: returns only paths updated after checkpoint
    - hash is always the hash of the FULL document
    """
    timestamp = next_timestamp()
    raw = await store.get_string(document_key)

    if not raw:
        return PullResult(data={}, hash="", timestamp=timestamp)

    parsed = json.loads(raw)
    doc = StoredDocument(
        v=parsed["v"],
        data=parsed["data"],
        timestamps=parsed["timestamps"],
        hash=parsed["hash"],
        author_pubkey=parsed.get("authorPubkey"),
        author_signature=parsed.get("authorSignature"),
    )

    if checkpoint and checkpoint > 0 and doc.timestamps:
        filtered = filter_by_checkpoint(doc.data, doc.timestamps, checkpoint)
        return PullResult(
            data=filtered,
            hash=doc.hash,
            timestamp=timestamp,
            author_pubkey=doc.author_pubkey,
            author_signature=doc.author_signature,
        )

    return PullResult(
        data=doc.data,
        hash=doc.hash,
        timestamp=timestamp,
        author_pubkey=doc.author_pubkey,
        author_signature=doc.author_signature,
    )

File: servers/ts/src/protocol/pull.ts
import type { IObjectStore } from "../interfaces.js"
import { nextTimestamp } from "../timestamp.js"
import type { StoredDocument, PullResult } from "./types.js"
import { filterByCheckpoint } from "./timestamps.js"

/**
 * Pull the current document, optionally filtered by checkpoint.
 * - No checkpoint (or 0): returns full data
 * - With checkpoint: returns only paths updated after checkpoint
 * - hash is always the hash of the FULL document
 *
 * @param store - Storage backend
 * @param documentKey - Unique key identifying this document in the store
 * @param checkpoint - Only return data updated after this timestamp
 */
export async function pull(
  store: IObjectStore,
  documentKey: string,
  checkpoint?: number,
): Promise<PullResult> {
  const timestamp = nextTimestamp()
  const raw = await store.getString(documentKey)

  if (!raw) {
    return { data: {}, hash: "", timestamp }
  }

  const doc = JSON.parse(raw) as StoredDocument

  if (checkpoint && checkpoint > 0 && Object.keys(doc.timestamps).length > 0) {
    const filtered = filterByCheckpoint(doc.data, doc.timestamps, checkpoint)
    return { data: filtered, hash: doc.hash, timestamp, authorPubkey: doc.authorPubkey, authorSignature: doc.authorSignature }
  }

  return { data: doc.data, hash: doc.hash, timestamp, authorPubkey: doc.authorPubkey, authorSignature: doc.authorSignature }
}
